src/main/scala/com/microsoft/hyperspace/index/rules/HyperspaceRule.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (2020) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.rules
+
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+
+import com.microsoft.hyperspace.ActiveSparkSession
+import com.microsoft.hyperspace.index.IndexLogEntry
+import com.microsoft.hyperspace.index.rules.ApplyHyperspace.PlanToIndexesMap
+
+/**
+ * Interface of exclusive type of indexes.
+ */
+trait HyperspaceRule extends ActiveSparkSession {
+
+  /**
+   * Sequence of conditions to apply indexes. Each filter contains conditions and
+   * filter out candidate indexes based on the conditions. The order of the sequence does matter
+   * because they are applied in order with the assumption that previous filter conditions met.
+   */
+  val filtersOnQueryPlan: Seq[QueryPlanIndexFilter]
+
+  /**
+   * Transform the plan to use the selected indexes after applying [[filtersOnQueryPlan]].
+   * All selected indexes should be able to be applied to the plan.
+   *
+   * @param plan Original query plan.
+   * @param indexes Selected indexes.
+   * @return Transformed plan to use the selected indexes.
+   */
+  def applyIndex(plan: LogicalPlan, indexes: Map[LogicalPlan, IndexLogEntry]): LogicalPlan
+
+  /**
+   * Calculate the score of the selected indexes.
+   *
+   * @param plan Original query plan.
+   * @param indexes Selected indexes.
+   * @return Score of selected indexes.
+   */
+  def score(plan: LogicalPlan, indexes: Map[LogicalPlan, IndexLogEntry]): Int
+
+  final def apply(plan: LogicalPlan, indexes: PlanToIndexesMap): (LogicalPlan, Int) = {
+    if (indexes.isEmpty) {
+      return (plan, 0)
+    }
+
+    val selectedIndexes = filtersOnQueryPlan
+      .foldLeft(indexes) { (pti, filter) =>
+        filter(plan, pti)
+      }
+      .map { kv =>
+        // After applying filtersOnQueryPlan, each candidate relation should have
+        // one candidate index.
+        assert(kv._2.length == 1)
```

Comment:
Should we enforce this with API for now, meaning `QueryPlanIndexFilter` returns `Map[LogicalPlan, IndexLogEntry]`?

Reply:
We could introduce additional filter which returns PlanToSelectedIndexes if necessary - `RankFilter`?

Reply:
yea seems fine to me. so, we will have:
```
all indexes ----> candidate indexes ---> selected indexes -----> best index
             |                       |                      |
      apply all SPIFs         apply all QPIFs       apply Ranker
```
, where
all indexes: all indexes available
candidate indexes: indexes after applying all SPIFs looking at only the source plan. These indexes may not be eligible against the overall query
selected indexes: index after applying all QPIFs looking at the overall query. Any of these selected indexes can be applied against the query, but they are not ranked.
best index: the best index to apply among selected indexes which is ranked by the Ranker.

Reply:
selected indexes sounds bit weird. how about applicableIndexes & best(or selected)Index ?